### PR TITLE
feat: log a warning when a relay fails to parse

### LINKED ISSUE #102: Log when a relay fails to parse
Relays must be provided in a particular configuration and if this fails for some reason the entry is silently discarded:

https://github.com/ralexstokes/mev-rs/blob/6f5bcb6ed8c2e4100f746ed15e6127ab04fa3f96/mev-boost-rs/src/service.rs#L56-L65

instead, we should log a warning when one relay fails to parse

### PR DESCRIPTION
**Pull request by chirag-bgh**
Closes #102

### CODE REVIEW

File: mev-boost-rs/src/service.rs [hunk 1]
```diff
@@ -53,12 +53,17 @@ pub struct Service {
...
+
+        for s in &config.relays {
+            match parse_url(s) {
+                Some(url) => match RelayEndpoint::try_from(url) {
+                    Ok(relay) => relays.push(relay),
+                    Err(_) => tracing::warn!("Failed to parse relay endpoint: {}", s),
+                },
+                None => tracing::warn!("Failed to parse URL: {}", s),
+            }
+        }
```
Comment (ralexstokes) [inline mev-boost-rs/src/service.rs:66]: 1. use more descriptive symbol for relay url than just `s`
2. and move this code to the `parse_url` fn and rename to reflect the expanded logic

```suggestion
        for relay_url in &config.relays {
            match input.parse::<Url>() {
                Ok(url) => match RelayEndpoint::try_from(url) {
                    Ok(relay) => relays.push(relay),
                    Err(err) => tracing::warn!("error parsing relay endpoint `{url}`: {err}"),
                },
                Err(err) => tracing::warn!("error parsing relay from URL `{relay_url}`: {err}"),
            }
        }
```

so `parse_url(input: &str) -> Option<Url>` becomes something like `parse_relay_endpoints(relay_urls: &[String]) -> Vec<RelayEndpoint>`

**Review by ralexstokes — COMMENTED** [review]
nice work! lets go one step further to clean up the `parse_url` fn and merge w/ this one

File: mev-boost-rs/src/service.rs [hunk 2]
```diff
@@ -30,18 +30,19 @@ impl Default for Config {
...
-    } else {
-        input
-            .parse()
-            .map_err(|err| {
-                tracing::warn!("error parsing relay from URL: `{err}`");
-                err
-            })
-            .ok()
+fn parse_relay_endpoints(relay_urls: &[String]) -> Vec<RelayEndpoint> {
+    let mut relays = Vec::new();
```
Comment (ralexstokes) [inline mev-boost-rs/src/service.rs:34]: nitpick but I prefer to just use the macro for this

```suggestion
    let mut relays = vec![];
```

File: mev-boost-rs/src/service.rs [hunk 3]
```diff
@@ -30,18 +30,19 @@ impl Default for Config {
...
+
+    for relay_url in relay_urls {
+        match relay_url.parse::<Url>() {
+            Ok(url) => match RelayEndpoint::try_from(url.clone()) {
```
Comment (ralexstokes) [inline mev-boost-rs/src/service.rs:38]: should be able to skip the clone here right?

```suggestion
            Ok(url) => match RelayEndpoint::try_from(url) {
```

**Review by ralexstokes — COMMENTED** [review]
nice refactor! two small things then we can merge

File: mev-boost-rs/src/service.rs (hunk 3 above)
Comment (chirag-bgh) [reply mev-boost-rs/src/service.rs:38]: No the compiler complains about this

Comment (ralexstokes) [reply mev-boost-rs/src/service.rs:38]: what does it complain about? this seems to be fine: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f36816a9416a02df4132432339f555bc

File: mev-boost-rs/src/service.rs [hunk 4]
```diff
@@ -30,18 +30,19 @@ impl Default for Config {
...
-            })
-            .ok()
+fn parse_relay_endpoints(relay_urls: &[String]) -> Vec<RelayEndpoint> {
+    let mut relays = vec![];
+
+    for relay_url in relay_urls {
+        match relay_url.parse::<Url>() {
+            Ok(url) => match RelayEndpoint::try_from(url.clone()) {
+                Ok(relay) => relays.push(relay),
+                Err(err) => tracing::warn!("error parsing relay endpoint `{url}`: {err}"),
```
Comment (chirag-bgh) [inline mev-boost-rs/src/service.rs:40]: If we don't clone then we would borrow a moved value here.

Comment (chirag-bgh) [reply mev-boost-rs/src/service.rs:40]: @ralexstokes

File: mev-boost-rs/src/service.rs [hunk 5]
```diff
@@ -30,18 +30,19 @@ impl Default for Config {
...
+fn parse_relay_endpoints(relay_urls: &[String]) -> Vec<RelayEndpoint> {
+    let mut relays = vec![];
+
+    for relay_url in relay_urls {
+        match relay_url.parse::<Url>() {
+            Ok(url) => match RelayEndpoint::try_from(url) {
+                Ok(relay) => relays.push(relay),
+                Err(err) => tracing::warn!("error parsing relay from URL `{relay_url}`: {err}"),
+            },
+            Err(err) => tracing::warn!("error parsing relay from URL `{relay_url}`: {err}"),
```
Comment (ralexstokes) [inline mev-boost-rs/src/service.rs:42]:
```suggestion
            Err(err) => tracing::warn!("error parsing relay URL `{relay_url}` from config: {err}"),
```

**Pull request merged**